x/swaprouter/router_test.go

```diff
@@ -33,45 +36,199 @@ func (suite *KeeperTestSuite) TestMultihopSwapExactAmountIn() {
+			routes: []types.SwapAmountInRoute{
+				{
+					PoolId:        1,
+					TokenOutDenom: "bar",
+				},
+				{
+					PoolId:        2,
+					TokenOutDenom: "baz",
+				},
+			},
+			incentivizedGauges:      []uint64{1, 2, 3, 4, 5, 6},
+			tokenIn:                 sdk.NewCoin("foo", sdk.NewInt(100000)),
+			tokenOutMinAmount:       sdk.NewInt(1),
+			expectReducedFeeApplied: false,
+		},
+		{
+			name: "Two routes: Swap between four asset pools - [foo -> uosmo](pool 1) - [uosmo -> baz](pool 2), with a half fee applied, both pools 1 percent fee",
+			poolCoins: []sdk.Coins{
+				sdk.NewCoins(sdk.NewCoin(bar, defaultInitPoolAmount), sdk.NewCoin(baz, defaultInitPoolAmount),
+					sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(uosmo, defaultInitPoolAmount)), // pool 1.
+				sdk.NewCoins(sdk.NewCoin(bar, defaultInitPoolAmount), sdk.NewCoin(baz, defaultInitPoolAmount),
+					sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(uosmo, defaultInitPoolAmount)), // pool 2.                                                                                     // pool 3.
+			},
+			poolFee: []sdk.Dec{defaultPoolSwapFee, defaultPoolSwapFee},
+			routes: []types.SwapAmountInRoute{
+				{
+					PoolId:        1,
+					TokenOutDenom: "uosmo",
+				},
+				{
+					PoolId:        2,
+					TokenOutDenom: "baz",
+				},
+			},
+			incentivizedGauges:      []uint64{1, 2, 3, 4, 5, 6},
+			tokenIn:                 sdk.NewCoin("foo", sdk.NewInt(100000)),
+			tokenOutMinAmount:       sdk.NewInt(1),
+			expectReducedFeeApplied: true,
+		},
+		{
+			name: "Three routes: Swap between four asset pools - [foo -> uosmo](pool 1) - [uosmo -> baz](pool 2) - [baz -> bar](pool 3), all pools 1 percent fee",
+			poolCoins: []sdk.Coins{
+				sdk.NewCoins(sdk.NewCoin(bar, defaultInitPoolAmount), sdk.NewCoin(baz, defaultInitPoolAmount),
+					sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(uosmo, defaultInitPoolAmount)), // pool 1.
+				sdk.NewCoins(sdk.NewCoin(bar, defaultInitPoolAmount), sdk.NewCoin(baz, defaultInitPoolAmount),
+					sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(uosmo, defaultInitPoolAmount)), // pool 2.
+				sdk.NewCoins(sdk.NewCoin(bar, defaultInitPoolAmount), sdk.NewCoin(baz, defaultInitPoolAmount),
+					sdk.NewCoin(foo, defaultInitPoolAmount), sdk.NewCoin(uosmo, defaultInitPoolAmount)), // pool 3.                                                                                      // pool 3.
+			},
+			poolFee: []sdk.Dec{defaultPoolSwapFee, defaultPoolSwapFee, defaultPoolSwapFee},
+			routes: []types.SwapAmountInRoute{
+				{
+					PoolId:        1,
+					TokenOutDenom: "uosmo",
+				},
+				{
+					PoolId:        2,
+					TokenOutDenom: "baz",
+				},
+				{
+					PoolId:        3,
+					TokenOutDenom: "bar",
+				},
+			},
+			incentivizedGauges:      []uint64{1, 2, 3, 4, 5, 6, 7, 8, 9},
+			tokenIn:                 sdk.NewCoin("foo", sdk.NewInt(100000)),
+			tokenOutMinAmount:       sdk.NewInt(1),
+			expectReducedFeeApplied: false,
```

Comment:
Did we have any tests for stableswap pools?

If not, could you add a TODO to add stableswap test cases please